evalutils/templates/evaluation/{{ cookiecutter.package_name }}/test.sh

```diff
@@ -1,21 +1,35 @@
 #!/usr/bin/env bash
 
-SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+# set -o errexit
 
-./build.sh
+# RANDOM_READ_BYTES is tha amount of bytes to read from /dev/urandom.
+# No particular reason for any specific value, just something vaguely sensible that provides a
+# finite limit on reading from /dev/urandom and gives the message digest (here hard-coded to md5)
+# something to calculate on...
+RANDOM_READ_BYTES=1024
+SCRIPTPATH="$(dirname "$(realpath "${0}")")"
 
-VOLUME_SUFFIX=$(dd if=/dev/urandom bs=32 count=1 | md5sum | cut --delimiter=' ' --fields=1)
+# echo "*** run ${SCRIPTPATH}/build.sh" >> "${TMPLOG}"
+# "${SCRIPTPATH}/build.sh"
 
-docker volume create {{ cookiecutter.package_name|lower }}-output-$VOLUME_SUFFIX
+PACKAGE_NAME={{ cookiecutter.package_name|lower }}
+VOLUME_SUFFIX=$(dd if=/dev/urandom bs=${RANDOM_READ_BYTES} count=1 | md5sum | cut --delimiter=' ' --fields=1)
+VOLUME_NAME="${PACKAGE_NAME}-output-${VOLUME_SUFFIX}"
 
+echo "*** vol name ${VOLUME_NAME} ; docker vol create: "
+docker volume create "${VOLUME_NAME}"
+
+echo "*** last exec: $? : docker run"
 docker run --rm \
         --memory=4g \
-        -v $SCRIPTPATH/test/:/input/ \
-        -v {{ cookiecutter.package_name|lower }}-output-$VOLUME_SUFFIX:/output/ \
-        {{ cookiecutter.package_name|lower }}
-
+        -v "${SCRIPTPATH/test/}":/input/ \
+        -v "${VOLUME_NAME}":/output/ \
+        "${PACKAGE_NAME}"   # cmd missing here? Or am I just mis-reading this?
```

Comment:
I don't think a cmd is missing, we just take the default that the container provides and never override it. Maybe @silvandeleemput can help out with this PR.

Reply:
So whatever is defined as `ENTRYPOINT` in the Dockerfile is run - usual thing?
The Dockerfile has this:
```
ENTRYPOINT "python" "-m" "evaluation"
```
From `...evaluation/{{...}}/test.sh` on the `master` branch I see this:
```bash
docker run --rm \
        --memory=4g \
        -v $SCRIPTPATH/test/:/input/ \
        -v {{ cookiecutter.package_name|lower }}-output-$VOLUME_SUFFIX:/output/ \
        {{ cookiecutter.package_name|lower }}
```

which runs a container with `tests` mounted at `/input` and the just created volume mounted at `/output`, whereby the image for the container should already have been built (`build.sh`).

The docker file gives this as the executable to be run:
```
COPY --chown=evaluator:evaluator evaluation.py /opt/evaluation/

ENTRYPOINT "python" "-m" "evaluation"
```
whereby `evaluation.py`  is actually `evaluation.py.j2` as a cookie-cutter template which should be instantiated to the plain `evalution.py`

Is as far as I've got with tracking things down.

Yes might be a good idea for Sil / someone from RSE to take a look, I already spent more time than I should on this instead of doing my PhD research...

Reply:
> So whatever is defined as `ENTRYPOINT` in the Dockerfile is run - usual thing? The Dockerfile has this:
> 
> ```
> ENTRYPOINT "python" "-m" "evaluation"
> ```
> 
Exactly, that's being run.


> From `...evaluation/{{...}}/test.sh` on the `master` branch I see this:
> 
> ```shell
> docker run --rm \
>         --memory=4g \
>         -v $SCRIPTPATH/test/:/input/ \
>         -v {{ cookiecutter.package_name|lower }}-output-$VOLUME_SUFFIX:/output/ \
>         {{ cookiecutter.package_name|lower }}
> ```
> 
> which runs a container with `tests` mounted at `/input` and the just created volume mounted at `/output`, whereby the image for the container should already have been built (`build.sh`).
> 
> The docker file gives this as the executable to be run:
> 
> ```
> COPY --chown=evaluator:evaluator evaluation.py /opt/evaluation/
> 
> ENTRYPOINT "python" "-m" "evaluation"
> ```
> 
> whereby `evaluation.py` is actually `evaluation.py.j2` as a cookie-cutter template which should be instantiated to the plain `evalution.py`
> 

That's correct as well, once an evaluation container is built it uses an instantiated jinja2 template.

I'll have a look at running the tests now to see what goes wrong.

Reply:
Ok I create a new branch with the needed fixes here: https://github.com/comic/evalutils/tree/323_fix_for_general_build_test_sh_improvements

@GabyRumc you can merge them into your branch if you want.
I identified three core issues:

1. must use: ```{{ cookiecutter.docker_base_container }}``` for all other operations than running the container and not `"${PACKAGE_NAME}"`.
2. `set -o errexit` in `build.sh` files causes failures (uncommented for now)
3.  `-v "${SCRIPTPATH/test/}":/input/ \` should be changed into `-v "${SCRIPTPATH}/test/":/input/ \` in the evalution test.sh file.

Reply:
> 1. must use: `{{ cookiecutter.docker_base_container }}` for all other operations than running the container and not `"${PACKAGE_NAME}"`.

Why? Once it is evaluated it remains constant.

> 2. `set -o errexit` in `build.sh` files causes failures (uncommented for now)

Aaargh! Disagree. This is entirely the point: none of the calls should fail.

> 3. `-v "${SCRIPTPATH/test/}":/input/ \` should be changed into `-v "${SCRIPTPATH}/test/":/input/ \` in the evalution test.sh file.

Not sure I see the difference between the two?

Reply:
> > 1. must use: `{{ cookiecutter.docker_base_container }}` for all other operations than running the container and not `"${PACKAGE_NAME}"`.
> 
> Why? Once it is evaluated it remains constant.

No, they are different docker images. So you have these two different images, which do different things: 
* `{{ cookiecutter.docker_base_container }}` -> running utils
* `{{ cookiecutter.package_name|lower }}` -> running the algorithm / evalution code

> 
> > 1. `set -o errexit` in `build.sh` files causes failures (uncommented for now)
> 
> Aaargh! Disagree. This is entirely the point: none of the calls should fail.
> 

I see your intention, however, for some reason, the docker build succeeds without a proper 0 return code there, and I don't know why yet. The image gets properly built though, and the rest of the process works. It would indeed be nice to know why it doesn't return the expected return code there despite functioning like expected.

> > 1. `-v "${SCRIPTPATH/test/}":/input/ \` should be changed into `-v "${SCRIPTPATH}/test/":/input/ \` in the evalution test.sh file.
> 
> Not sure I see the difference between the two?

```bash
#!/usr/bin/env bash

SCRIPTPATH="blah"
echo "${SCRIPTPATH/test/}"
echo "${SCRIPTPATH}/test/"

```

output:

```
blah
blah/test/
```

Reply:
> 
> No, they are different docker images. So you have these two different images, which do different things:
> 
> * `{{ cookiecutter.docker_base_container }}` -> running utils
> * `{{ cookiecutter.package_name|lower }}` -> running the algorithm / evalution code
> 

+1 will re-check. Still doesn't explain the issues I had even when I didn't re-factor the cookiecutter stuff though, but for sure no need to add problems to those already there ;)

> > > 1. `set -o errexit` in `build.sh` files causes failures (uncommented for now)
> for some reason, the docker build succeeds without a proper 0 return code there, and I don't know why yet.

which is precisely the point :-)

> 
> > > 1. `-v "${SCRIPTPATH/test/}":/input/ \` should be changed into `-v "${SCRIPTPATH}/test/":/input/ \` in the evalution test.sh file.
> > 
> > 
> > Not sure I see the difference between the two?
> 

Saw it: simple typo with the `}` in the wrong place. Nice catch :-) one of those you could stare at for ages and not see it, `shellcheck` did not catch this one either as syntactically correct, although not semantically correct

Reply:
> Ok I create a new branch with the needed fixes here: 
> https://github.com/comic/evalutils/tree/323_fix_for_general_build_test_sh_improvements
> @GabyRumc you can merge them into your branch if you want.

Will do, thx - all except `set -o errexit`, which really should stay (although for debugging purposes only I clearly needed to comment it out at times)

Reply:
That

> > Ok I create a new branch with the needed fixes here:
> > https://github.com/comic/evalutils/tree/323_fix_for_general_build_test_sh_improvements
> > @GabyRumc you can merge them into your branch if you want.
> 
> Will do, thx - all except `set -o errexit`, which really should stay (although for debugging purposes only I clearly needed to comment it out at times)

No problem. That's fine as well, you might still get errors then, but at least you should be able to narrow it down now.

Reply:
In the end, i just did a manual port: only 2 edits to test.sh (one of which was a simple typo re `}`)
`set -o errexit` off for now, am trying to run all template tests first, takes some time
Once they pass successfully, I'll take another look at the docker error code issue.
@silvandeleemput thx for the review -> a second pair of eyes almost always helps :-)

Reply:
Oh, one other thing I forgot to mention: in the algorithm `test.sh` you should make sure that the output file is not a metrics.json file but a results.json file instead (the algorithm template is significantly different from the evaluation, with different outputs):

So:
 `{{ cookiecutter.docker_base_container }} cat /output/metrics.json | python -m json.tool` 
should become: 
 `{{ cookiecutter.docker_base_container }} cat /output/results.json | python -m json.tool`

Otherwise, some tests are likely to fail.

Reply:
Even though `results.json` is deprecated according to the documentation at https://grand-challenge.org/algorithms/interfaces/ ?
![image](https://user-images.githubusercontent.com/61788529/146923971-c51f11c0-faa2-482f-8ee3-61c1607ecc3c.png)

Reply:
Ah, yes, for now, we do, since the algorithm template will generate a `results.json`. But you are totally right, it is outdated/deprecated to do it this way, I think the results.json should not be generated by the generated algorithm template anymore. It should probably be removed and instead, it should generate whatever the user wants to generate as output by default. We should make a new issue for this.

Reply:
Agree with new issue, so for now I'll just revert to `results.json` in `algorithm...test.sh` just to get the code running cleanly as-is